trace-recorder/src/main/org/jetbrains/lincheck/trace/recorder/TraceCollectingEventTracker.kt

```diff
@@ -906,45 +925,73 @@ class TraceCollectingEventTracker(
         )
         strategy.registerCurrentThread(threadData.threadId)
         strategy.tracePointCreated(null,tracePoint)
+        threadData.setRootCall(tracePoint)
         threadData.pushStackFrame(tracePoint, null, isInline = false)
 
         startTime = System.currentTimeMillis()
     }
 
-    fun completeRunningThread(thread: Thread, threadDescriptor: ThreadDescriptor) {
+    private fun completeRunningThread(thread: Thread, threadDescriptor: ThreadDescriptor) {
         require(!threadDescriptor.isAnalysisEnabled) { "When completing a Thread ${thread.name} (${thread.id}), its analysis is expected to be disabled" }
         val threadData = threadDescriptor.eventTrackerData as? ThreadData? ?: return
         // wait until that thread finishes whatever injected code it is executing right now
         spinner.spinWaitUntil { !threadDescriptor.isInsideInjectedCode }
         // now, we are sure that another thread has finished its injected code
         // and will not attempt to execute anything else, because we disabled analysis in it
-        val stack = threadData.getStack()
-        while (stack.isNotEmpty()) {
-            val frame = stack.last()
-            val tracePoint = frame.call
+
+        // Skip thread completely if it is finished "naturally"
+        // Early exit because we must skip `strategy.completeThread(thread)`
+        // for this thread too.
+        if (threadData.getStack().isEmpty()) return
+
+        while (threadData.getStack().isNotEmpty()) {
             // complete all open loops with their last iterations
             while (threadData.currentLoopTracePoint() != null) {
                 exitCurrentLoop(thread, threadData)
             }
+            // Pop stack
+            val tracePoint = threadData.popStackFrame()
             // complete the method call
             tracePoint.result = TR_OBJECT_UNFINISHED_METHOD_RESULT
             strategy.completeContainerTracePoint(thread, tracePoint)
-            // do not pop the last method call in order for 'text' tracing mode to work
-            // it is used for debugging in the `TRPlayground.kt` file
-            if (stack.size != 1) {
-                threadData.popStackFrame() // remove the current frame from the stack
-            } else {
-                // basically, the last `Thread.run()` call is preserved as a root node
-                // for 'text' tracing mode to work properly
-                break
+        }
```

Comment:
The logic inside this while loop could be united with the `forceThreadEnd` method. 
They only differentiate in assigning method call result/exception and loops handling